cmd/agent-smoke/internal/smoke.go

```diff
@@ -0,0 +1,158 @@
+package smoke
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"golang.org/x/sync/errgroup"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+// Smoke is the top level object for a smoke test.
+type Smoke struct {
+	logger     log.Logger
+	clientset  *kubernetes.Clientset
+	kubeconfig string
+	namespace  string
+	tasks      []Task
+
+	chaosFrequency    time.Duration
+	mutationFrequency time.Duration
+}
+
+// NewSmokeTest is the constructor for a Smoke object.
+func NewSmokeTest(opts ...Option) (*Smoke, error) {
```

Comment:
stylistic suggestion, I don't care if you actually make this change: 

I would personally cut down on boilerplate until there's a reason it can't be avoided. I think the options pattern is neat and useful if your options need to do logic, but right now you don't need that and it's bubbling throughout the code. 

As an alternative, if you had an options strut:

```go
type Options struct {
  Kubeconfig string 
  Namespace string 
  MutationFrequency, ChaosFrequency time.Duration 
}
```

then you could avoid the boilerplate to define/call the options and even bind flags to field members directly in the main package:

```go 
package main 

func main() {
  var (
     ll logging.Level 
     o Options 
  )
  
  ll.RegisterFlags(flag.CommandLine) 
  flag.StringVar(&o.Namespace, "namespace", "agent-smoke-test", "namespace smoke test should run in")
  flag.StringVar(&o.Kubeconfig, "kubeconfig", "", "absolute path to the kubeconfig file")
  // etc 
  
  logger := util.NewLogger(ll) 
  st, err := smoke.NewSmokeTest(logger, o) 
  if err != nil {
     // ...
  } 
}
```

Reply:
I initially thought some of the options might require some logic when I started coding on this but that didn't end up being the case. I got rid of the functional options and the code is definitely less clunky.